internal/datastore/proxy/context.go

```diff
@@ -0,0 +1,119 @@
+	return DatastoreFromContext(ctx).DeleteRelationships(ctx, preconditions, filter)
+}
+
+func (cd ctxDatastore) WriteTuples(ctx context.Context, preconditions []*v1.Precondition, updates []*v1.RelationshipUpdate) (datastore.Revision, error) {
+	return DatastoreFromContext(ctx).WriteTuples(ctx, preconditions, updates)
+}
+
+func (cd ctxDatastore) Revision(ctx context.Context) (datastore.Revision, error) {
+	return DatastoreFromContext(ctx).Revision(ctx)
+}
+
+func (cd ctxDatastore) SyncRevision(ctx context.Context) (datastore.Revision, error) {
+	return DatastoreFromContext(ctx).SyncRevision(ctx)
+}
+
+func (cd ctxDatastore) Watch(ctx context.Context, afterRevision datastore.Revision) (<-chan *datastore.RevisionChanges, <-chan error) {
+	return DatastoreFromContext(ctx).Watch(ctx, afterRevision)
+}
+
+func (cd ctxDatastore) WriteNamespace(ctx context.Context, newConfig *v0.NamespaceDefinition) (datastore.Revision, error) {
+	return DatastoreFromContext(ctx).WriteNamespace(ctx, newConfig)
+}
+
+func (cd ctxDatastore) ReadNamespace(ctx context.Context, nsName string) (*v0.NamespaceDefinition, datastore.Revision, error) {
+	return DatastoreFromContext(ctx).ReadNamespace(ctx, nsName)
+}
+
+func (cd ctxDatastore) DeleteNamespace(ctx context.Context, nsName string) (datastore.Revision, error) {
+	return DatastoreFromContext(ctx).DeleteNamespace(ctx, nsName)
+}
+
+func (cd ctxDatastore) QueryTuples(ctx context.Context, filter datastore.TupleQueryResourceFilter, revision datastore.Revision) datastore.TupleQuery {
+	return DatastoreFromContext(ctx).QueryTuples(ctx, filter, revision)
+}
+
+func (cd ctxDatastore) ReverseQueryTuplesFromSubjectNamespace(ctx context.Context, subjectNamespace string, revision datastore.Revision) datastore.ReverseTupleQuery {
+	return DatastoreFromContext(ctx).ReverseQueryTuplesFromSubjectNamespace(ctx, subjectNamespace, revision)
+}
+
+func (cd ctxDatastore) ReverseQueryTuplesFromSubject(ctx context.Context, subject *v0.ObjectAndRelation, revision datastore.Revision) datastore.ReverseTupleQuery {
+	return DatastoreFromContext(ctx).ReverseQueryTuplesFromSubject(ctx, subject, revision)
+}
+
+func (cd ctxDatastore) ReverseQueryTuplesFromSubjectRelation(ctx context.Context, subjectNamespace, subjectRelation string, revision datastore.Revision) datastore.ReverseTupleQuery {
+	return DatastoreFromContext(ctx).ReverseQueryTuplesFromSubjectRelation(ctx, subjectNamespace, subjectRelation, revision)
+}
+
+func (cd ctxDatastore) CheckRevision(ctx context.Context, revision datastore.Revision) error {
+	return DatastoreFromContext(ctx).CheckRevision(ctx, revision)
+}
+
+func (cd ctxDatastore) ListNamespaces(ctx context.Context) ([]*v0.NamespaceDefinition, error) {
+	return DatastoreFromContext(ctx).ListNamespaces(ctx)
+}
+
+// Create a new type to prevent context collisions
+type datastoreKey string
+
+var datastoreCtxKey datastoreKey = "injected-datastore"
+
+type dsHandle struct{ ds datastore.Datastore }
```

Comment:
Why do we need a handle and not just the datastore itself?

Reply:
Its a pattern we follow everywhere, in case we need to add additional stuff to the context

Reply:
It's really easy to add it in the future when we need it rather than speculatively now. We don't do this in that many places and I consistently advocate against it.

Reply:
Okay